internal/addoninstancesdk/sdk.go

```diff
@@ -0,0 +1,152 @@
+	AddonTargetNamespace    string
+
+	// to "concurrent-safely" track whether there's a heartbeat reporter running or not
+	isRunning      bool
+	isRunningMutex *sync.Mutex
+
+	// the latest condition which the heartbeat reporter would be reporting the periodically
+	latestCondition metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	currentInterval time.Duration
+	ticker          *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	stopperCh chan bool
+	updateCh  chan updateOptions
+}
+
+// ensure that the `AddonInstanceHeartbeatReporter` implements the `AddonInstanceStatusReporterClient` interface
+var _ AddonInstanceStatusReporterClient = (*AddonInstanceHeartbeatReporter)(nil)
+
+// InitializeAddonInstanceHeartbeatReporterSingleton sets up a singleton of the type `AddonInstanceHeartbeatReporter` (only if it doesn't exist yet) and returns it to the caller.
+func InitializeAddonInstanceHeartbeatReporterSingleton(addonInstanceInteractor AddonInstanceInteractorClient, addonName string, addonTargetNamespace string) (*AddonInstanceHeartbeatReporter, error) {
+	if addonInstanceHeartbeatReporterSingleton == nil {
+		addonInstanceHeartbeatReporterSingletonMutex.Lock()
+		defer addonInstanceHeartbeatReporterSingletonMutex.Unlock()
+		if addonInstanceHeartbeatReporterSingleton == nil {
+			addonInstanceHeartbeatReporterSingleton = &AddonInstanceHeartbeatReporter{
+				AddonInstanceInteractor: addonInstanceInteractor,
+				AddonName:               addonName,
+				AddonTargetNamespace:    addonTargetNamespace,
+				isRunning:               false,
+				isRunningMutex:          &sync.Mutex{},
+				latestCondition: metav1.Condition{
+					Type:    "addons.managed.openshift.io/Healthy",
+					Status:  "False",
+					Reason:  "NoHeartbeatReported",
+					Message: fmt.Sprintf("Addon '%s' hasn't reported any heartbeat yet", addonName),
+				},
+				stopperCh: make(chan bool),
+				updateCh:  make(chan updateOptions),
+			}
+
+			currentAddonInstance, err := addonInstanceHeartbeatReporterSingleton.AddonInstanceInteractor.GetAddonInstance(context.TODO(), "addon-instance", addonInstanceHeartbeatReporterSingleton.AddonTargetNamespace)
+			if err != nil {
+				return nil, fmt.Errorf("error occurred while fetching the current heartbeat update period interval")
+			}
+			addonInstanceHeartbeatReporterSingleton.currentInterval = currentAddonInstance.Spec.HeartbeatUpdatePeriod.Duration
+		}
+	}
+
+	return addonInstanceHeartbeatReporterSingleton, nil
+}
+
+func GetAddonInstanceHeartbeatReporterSingleton() (*AddonInstanceHeartbeatReporter, error) {
+	if addonInstanceHeartbeatReporterSingleton == nil {
+		return nil, fmt.Errorf("heartbeat-reporter not found to be initialised. Initialize it by calling `InitializeAddonInstanceHeartbeatReporterSingleton(...)`")
+	}
+	return addonInstanceHeartbeatReporterSingleton, nil
+}
+
+func (adihrClient *AddonInstanceHeartbeatReporter) changeRunningState(desiredState bool) {
```

Comment:
>  create a new instance when you want to restart a StatusReporter during runtime.

In my opinion, this rather sounds to introduce more overhead and complexity because the restart would go all the way from the stripping the existing StatusReporter object and bootstrap a new one with essentially the same config. 

From my pov, right now in the PR, the stop/start aren't doing much except toggling the `StatusReporter.isRunning` attribute from the perspective of StatusReporter alongside safely switching on/off the heartbeat-reporter loop.

For a client/tenant, Status Reporter Restarts are going to be as simple as just calling `Stop()` followed by `Start()` so doesn't something they'd require from the SDK. But still if we wanna ship a dedicated `Restart()` method, it would be as simple as

```go
func (sr *StatusReporter) Restart(ctx context.Context) error {
    if err := sr.Stop(); err != nil {
        return err
    }
    if err := sr.Start(ctx); err != nil {
        return err
    }
}
```

If you're worried about a situation where the client/tenant would actually want to stop AND DESTROY the StatusReporter singleton object (if they're nitpicky about the associated memory allocations), then we can choose to ship a 1-liner `Delete()` or `Destroy()` but I find that highly unnecessary and exposing the SDK to multiple-edge cases.